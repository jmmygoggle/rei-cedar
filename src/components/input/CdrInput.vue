<template>
  <cdr-label-standalone
    :for-id="id"
    :label="label"
    :hide-label="hideLabel"
    :required="required"
    :optional="optional"
    :disabled="disabled"
  >
    <template
      slot="helper"
      v-if="hasHelperTop"
    >
      <slot name="helper-text-top" />
    </template>
    <template
      slot="info"
      v-if="hasInfo"
    >
      <slot name="info" />
    </template>
  </cdr-label-standalone>
  <div :class="$style['cdr-input-outer-wrap']">
    <div :class="[$style['cdr-input-wrap'], $style[focusedClass]]">
      <textarea
        v-if="rows && rows > 1"
        :rows="rows"
        :class="[
          $style[baseClass],
          $style[multilineClass],
          $style[preIconClass],
          $style[postIconClass],
          $style[postIconsClass],
          $style[errorClass],
          $style[backgroundClass],
          $style[sizeClass],
        ]"
        :id="id"
        :disabled="disabled"
        :required="required"
        :aria-label="hideLabel ? label : null"
        v-bind="$attrs"
        :value="modelValue"
        @input="$emit('update:modelValue', $event.target.value)"
        @focus="isFocused = true"
        @blur="isFocused = false"
      />
      <input
        v-else
        type="type"
        :class="[
          $style[baseClass],
          $style[preIconClass],
          $style[postIconClass],
          $style[postIconsClass],
          $style[errorClass],
          $style[backgroundClass],
          $style[sizeClass],
        ]"
        :id="id"
        :disabled="disabled"
        :required="required"
        :aria-label="hideLabel ? label : null"
        v-bind="$attrs"
        :value="modelValue"
        @input="$emit('update:modelValue', $event.target.value)"
        @focus="isFocused = true"
        @blur="isFocused = false"
      >
      <span
        v-if="hasPreIcon"
        :class="$style['cdr-input__pre-icon']"
      >
        <slot name="pre-icon" />
      </span>

      <span
        v-if="hasPostIcon"
        :class="$style['cdr-input__post-icon']"
      >
        <slot name="post-icon" />
      </span>
    </div>
    <div
      v-if="hasInfoAction"
      :class="$style['cdr-input__info-action']"
    >
      <slot name="info-action" />
    </div>
  </div>
  <span
    :class="$style['cdr-input__helper-text']"
    v-if="hasHelperBottom && !error"
  >
    <slot name="helper-text-bottom" />
  </span>
  <cdr-form-error
    error="{this.error}"
    v-if="error"
  >
    <template slot="error">
      <slot name="error" />
    </template>
  </cdr-form-error>
</template>
<script lang="ts">
import { defineComponent, computed, ref } from 'vue';
import propValidator from '../../utils/propValidator';
import CdrLabelStandalone from '../labelStandalone/CdrLabelStandalone';
import CdrFormError from '../formError/CdrFormError';
import sizeProps from '../../props/size';
import backgroundProps from '../../props/background';
import { buildClass } from '../../utils/buildClass';

// CH CH CHANGES:
// ID is now required !!!
// HELPER-TEXT is now helper-text-bottom
// VALUE is modeelValue
// WRAPPER ELEMENT REMOVED! WRAP IF YOU NEED TO!

// CHECK THESE THINGS:
// does it still calculate postIconSSSSSS correctly?
// slots passed into sub-components working?
// do we need to re-emit focus/blur?!?!?!

export default defineComponent({
  name: 'CdrInput',
  components: {
    CdrLabelStandalone,
    CdrFormError,
  },
  inheritAttrs: false,
  props: {
    /**
     * `id` for the input that is mapped to the label `for` attribute.
    */
    id: {
      type: String,
      required: true,
    },
    /**
     *  'type' attribute for the input as defined by w3c.
     *  Only supporting text|email|number|password|search|url.
     *  The increment/decrement webkit psuedo element is hidden for number.
    */
    type: {
      type: [String],
      default: 'text',
      validator: (value) => propValidator(
        value,
        // TODO: moar types?
        ['text', 'email', 'number', 'password', 'search', 'url', 'tel'],
      ),
    },
    /**
     * Label text. This is required for a11y even if hiding the label with `hideLabel`.
    */
    label: {
      type: String,
      required: true,
    },
    /**
     * Removes the label element but sets the input `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Number of rows for input.  Converts component to text-area if rows greater than 1.
    */
    rows: Number,
    // Set which background type the input renders on
    background: backgroundProps,
    size: sizeProps,
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    disabled: Boolean,
    required: Boolean,
    optional: Boolean,
    modelValue: {
      type: [String, Number],
    },
  },
  setup(props, ctx) {
    const baseClass = 'cdr-input';

    const isFocused = ref(false);
    const hasHelperTop = ctx.slots['helper-text-top'];
    const hasHelperBottom = ctx.slots['helper-text-bottom'];
    const hasPreIcon = ctx.slots['pre-icon'];
    const hasPostIcon = ctx.slots['post-icon'];
    const hasPostIcons = hasPostIcon && ctx.slots['post-icon'].length > 1;
    const hasInfo = ctx.slots.info;
    const hasInfoAction = ctx.slots['info-action'];

    const multilineClass = computed(() => props.rows > 1 && 'cdr-input--multiline');
    const preIconClass = computed(() => hasPreIcon && 'cdr-input--preicon');
    const postIconClass = computed(() => hasPostIcon && 'cdr-input--posticon');
    const postIconsClass = computed(() => hasPostIcons && 'cdr-input--posticons');
    const errorClass = computed(() => props.error && 'cdr-input--error');
    const backgroundClass = computed(() => `cdr-input--${props.background}`);
    const sizeClass = computed(() => props.size && buildClass(baseClass, props.size));
    const focusedClass = computed(() => isFocused.value && 'cdr-input--focus');

    return {
      baseClass,
      sizeClass,
      focusedClass,
      multilineClass,
      preIconClass,
      postIconClass,
      postIconsClass,
      errorClass,
      backgroundClass,
      isFocused,
      hasHelperTop,
      hasHelperBottom,
      hasPreIcon,
      hasPostIcon,
      hasInfo,
      hasInfoAction,
    };
  },
});
</script>
<style lang="scss" module src="./styles/CdrInput.scss">
</style>
