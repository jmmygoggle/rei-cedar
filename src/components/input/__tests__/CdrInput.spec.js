import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import sinon from 'sinon';
import CdrInput from 'componentdir/input/CdrInput';

describe('CdrInput', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        label: 'Label Test',
        id: 'renders'
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders number input correctly', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        label: 'Label Test',
        id: 'renders',
        type: 'number',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('generates an id correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'testing',
      },
    });
    expect(wrapper.vm.$refs.input.id).toBe(wrapper.vm._uid.toString());
  });

  it('renders an input element', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'Label Test',
      },
    });
    expect(wrapper.find('input').element.tagName).toBe('INPUT');
  });

  it('sets input name attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'Label Test',
      },
      attrs: {
        name: 'testName',
      }
    });
    expect(wrapper.find('input').attributes('name')).toBe('testName');
  });

  it('sets input disabled attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        disabled: true,
      },
    });
    expect(wrapper.find('input').attributes('disabled')).toBe(true);
  });

  it('sets input readonly attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      attrs: {
        readonly: true,
      },
    });
    expect(wrapper.find('input').attributes('readonly')).toBe(true);
  });

  it('sets input required attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        required: true,
      },
    });
    expect(wrapper.find('input').attributes('required')).toBe(true);
  });

  it('sets attrs for numeric inputt', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        required: true,
        type: 'number'
      },
    });
    expect(wrapper.vm.$refs.input.hasAttribute('novalidate')).toBe(true);
    expect(wrapper.vm.$refs.input.hasAttribute('pattern')).toBe(true);
    expect(wrapper.vm.$refs.input.hasAttribute('inputmode')).toBe(true);
  });

  it('sets input autofocus attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      attrs: {
        autofocus: true,
      },
    });
    expect(wrapper.find('input').attributes('autofocus')).toBe(true);
  });

  it('sets input maxlength attribute correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      attrs: {
        maxlength: '20',
      },
    });
    expect(wrapper.find('input').attributes('maxlength')).toBe('20');
  });

  it('sets input placeholder correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      attrs: {
        placeholder: 'test placeholder',
      },
    });
    expect(wrapper.find('input').attributes('placeholder')).toBe('test placeholder');
  });

  it('sets multiline to textarea correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        rows: 2,
      },
    });
    expect(wrapper.find('textarea').element.tagName).toBe('TEXTAREA');
  });

  it('sets multiline rows correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        multiLine: true,
        rows: 10
      },
    });
    expect(wrapper.find('textarea').attributes('rows')).toBe('10');
  });

  it('sets default input type correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
    });
    expect(wrapper.find('input').attributes('type', 'text')).toBe(true);
  });

  it('overrides input type correctly', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        type: 'url',
      },
    });
    expect(wrapper.find('input').attributes('type', 'url')).toBe(true);
  });

  it('emits an input event', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        value: 'bar'
      },
    });
    const input = wrapper.find('input');
    input.setValue('foo');
    expect(wrapper.emitted().input[0][0]).toBe('foo');
  });

  it('emits a blur event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      listeners: {
        'blur': spy
      }
    });
    const input = wrapper.find('input');
    input.trigger('blur')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('emits a focus event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      listeners: {
        'focus': spy
      }
    });
    const input = wrapper.find('input');
    input.trigger('focus')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('adds focused class to wrapper on input focus', async () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      listeners: {
        'focus': spy
      }
    });
    const input = wrapper.find('input');
    input.trigger('focus')
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.cdr-input--focus').exists()).toBeTruthy();
    input.trigger('blur')
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.cdr-input--focus').exists()).toBeFalsy();
  });

  it('emits a paste event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      listeners: {
        'paste': spy
      }
    });
    const input = wrapper.find('input');
    input.trigger('paste')
    expect(spy.calledOnce).toBeTruthy();
  });

  it('emits a keydown event', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test'
      },
      listeners: {
        'keydown': spy
      }
    });
    const input = wrapper.find('input');
    input.trigger('keydown', {
      key: 'a'
    })
    expect(spy.called).toBeTruthy();
  });

  it('renders helper-text-bottom slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      slots: {
        'helper-text-bottom': 'very helpful',
      },
    });
    expect(wrapper.find('.cdr-input__helper-text').text()).toBe('very helpful');
  });

  it('renders pre-icon slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      slots: {
        'pre-icon': 'ðŸ¤ ',
      },
    });
    expect(wrapper.find('.cdr-input__pre-icon').text()).toBe('ðŸ¤ ');
  });

  it('renders post-icon slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      slots: {
        'post-icon': 'ðŸ˜Ž',
      },
    });
    expect(wrapper.find('.cdr-input__post-icon').text()).toBe('ðŸ˜Ž');
  });

  it('adds spacing class when post-icon slot is present', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      slots: {
        'post-icon': 'ðŸ˜Ž',
      },
    });
    expect(wrapper.find('.cdr-input--posticon').exists()).toBe(true);
  });


  it('adds spacing class when multiple elements are present in post-icon slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
      },
      slots: {
        'post-icon': '<div>ðŸ˜Ž</div><div>ðŸ¤ </div>',
      },
    });
    expect(wrapper.find('.cdr-input--posticons').exists()).toBe(true);
  });

  it('renders info action slot', () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        label: 'test',
        id: 'info-action'
      },
      slots: {
        'info-action': 'ðŸ¤ ',
      },
    });
    expect(wrapper.find('.cdr-input__info-action').text()).toBe('ðŸ¤ ');
  });

  it('renders error slot when error state is active', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        error: true
      },
      slots: {
        'error': 'whoops',
      },
    });
    expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
  });

  it('renders text when passed as error', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        error: 'incorrect!'
      },
    });
    expect(wrapper.find('.cdr-form-error').text()).toBe('incorrect!');
  });

  it('does not render error slot when error state is inactive', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        error: false
      },
      slots: {
        'error': 'whoops',
      },
    });
    expect(wrapper.find('.cdr-form-error').exists()).toBe(false);
  });

  it('renders error slot instead of bottom helper slot when error is active', () => {
    const wrapper = mount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        error: true,
      },
      slots: {
        'error': 'whoops',
        'helper-text-bottom': 'not me'
      },
    });
    expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
    expect(wrapper.find('.cdr-input__helper-text-bottom').exists()).toBe(false);
  });

  // NOTE - can't use v-model directly here, targeting the `data` prop instead
  it('updating v-model data updates the input', async () => {
    const wrapper = shallowMount(CdrInput, {
      propsData: {
        id: 'test',
        label: 'test',
        value: 'bar'
      },
    });
    const input = wrapper.find('input');
    wrapper.setProps({value: ''});
    await wrapper.vm.$nextTick();
    expect(input.element.value).toBe('');
  });
});
