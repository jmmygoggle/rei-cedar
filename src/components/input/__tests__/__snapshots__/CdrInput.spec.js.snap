// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrInput renders correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test
    </label>
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary"
        id="renders"
        spellcheck="false"
        type="text"
      />
    </div>
  </div>
  <div
    class="cdr-label-standalone__post-content"
  />
</div>
`;

exports[`CdrInput renders error state correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test
    </label>
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    <div
      class="cdr-input-wrap"
    >
      <input
        aria-errormessage="renders-error"
        aria-invalid="true"
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--error cdr-input--primary"
        id="renders"
        spellcheck="false"
        type="text"
      />
    </div>
  </div>
  <div
    class="cdr-label-standalone__post-content"
  >
    <div
      class="cdr-form-error"
      id="renders-error"
      role="status"
      tabindex="0"
    >
      <span
        class="cdr-form-error__icon"
      />
       Something is wrong!
    </div>
  </div>
</div>
`;

exports[`CdrInput renders number input correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test
    </label>
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary"
        id="renders"
        inputmode="numeric"
        novalidate="novalidate"
        pattern="[0-9]*"
        spellcheck="false"
        type="number"
      />
    </div>
  </div>
  <div
    class="cdr-label-standalone__post-content"
  />
</div>
`;
